File: 2022/07/7.py
import re

cmds = open("input.txt").read().splitlines()

folders = {}
path = ()

for cmd in cmds:
  if cmd == "$ cd ..":
    path = path[:-1]
  elif cmd.startswith("$ cd"):
    path = (*path, cmd.split(" ")[-1])
  elif re.match(r"^\d+", cmd):
    size = int(cmd.split(" ")[0])
    for i in range(len(path), 0, -1):
      folders[path[:i]] = size + folders.get(path[:i], 0)

print("Part 1:", sum(f for f in folders.values() if f <= 100000))
print("Part 2:", min(f for f in folders.values() if f >= 30000000 - (70000000 - folders[("/",)])))

File: 2022/07/7.clj
(ns aoc.2022.07.7
  (:require [aoc.common :refer [read-input sum]]
            [clojure.string :as str]))

(defn build-tree
  ([cmds] (build-tree cmds {} []))
  ([cmds tree path]
   (if-let [cmd (first cmds)]
     (let [cmds (rest cmds) parts (str/split cmd #" ")]
       (condp re-matches cmd
         #"\$ cd \.\." (recur cmds tree (pop path))
         #"\$ cd .+"   (recur cmds tree (conj path (last parts)))
         #"\d+ .+"     (recur cmds (assoc-in tree (conj path :files (last parts)) (parse-long (first parts))) path)
         (recur cmds tree path)))
     tree)))

(defn folder-sizes
  ([tree] (folder-sizes tree {} []))
  ([tree folders path]
   (let [dirs (->> (get-in tree path) keys (filter #(not= :files %)))]
     (->> dirs
          (reduce (fn [folders f]
                    (let [p (conj path f)
                          filesize (sum (vals (get-in tree (conj p :files))))
                          folders (folder-sizes tree folders p)
                          subfolders (->> folders (filter #(= p (pop (first %)))) vals sum)]
                      (assoc folders p (+ filesize subfolders))))
                  folders)))))

(let [folders (->> (read-input)
                   build-tree
                   folder-sizes)
      free-space (- 70000000 (get folders ["/"]))]

  (->> folders
       vals
       (filter #(<= % 100000))
       sum
       (println "Part 1:"))

  (->> folders
       vals
       (filter #(>= % (- 30000000 free-space)))
       (apply min)
       (println "Part 2:")))